### PR TITLE
Refactor get identity provider factory

### PR DESCRIPTION
**Pull request by tzheleznyak**
Refactor factory to identity provbider to be without constructor

### CODE REVIEW

File: app/domain/authentication/authn_jwt/configuration_jwt_generic_vendor.rb [hunk 1]
```diff
@@ -3,7 +3,7 @@ module AuthnJwt
     # Mock JWTConfiguration class to use it to develop other part in the jwt authenticator
     class ConfigurationJWTGenericVendor < ConfigurationInterface
       def self.jwt_identity(authentication_parameters)
-        id_provider = Authentication::AuthnJwt::IdentityProviderFactory.new(authentication_parameters).relevant_id_provider
+        id_provider = Authentication::AuthnJwt::IdentityProviderFactory.create_identity_provider(authentication_parameters: authentication_parameters)
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/configuration_jwt_generic_vendor.rb:6]:
```suggestion
        id_provider = Authentication::AuthnJwt::IdentityProviderFactory
  .create_identity_provider(
    authentication_parameters: authentication_parameters
  )
```

File: app/domain/authentication/authn_jwt/identity_provider_factory.rb [hunk 2]
```diff
@@ -5,20 +5,23 @@ module AuthnJwt
...
-      end
+      def self.create_identity_provider(
+        authentication_parameters:,
+        identity_from_url_provider_class: Authentication::AuthnJwt::IdentityFromUrlProvider,
+        identity_from_decoded_token_class: Authentication::AuthnJwt::IdentityFromDecodedTokenProvider
+      )
+        identity_provider = identity_from_decoded_token_class.new(authentication_parameters)
+        puts "decoded token identity avalible? = #{identity_provider.identity_available?}"
+        if identity_provider.identity_available?
+          Rails.logger.debug(LogMessages::Authentication::AuthnJwt::DECODED_TOKEN_IDENTITY_PROVIDER_SELECTED.new)
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/identity_provider_factory.rb:16]: `Rails.logger` should also be a dependency (i.e param with default value in method signature)

File: app/domain/authentication/authn_jwt/identity_provider_factory.rb [hunk 3]
```diff
@@ -5,20 +5,23 @@ module AuthnJwt
...
+          return identity_provider
+        end
 
-      def relevant_id_provider
-        if @from_decoded_token_provider.identity_available?
+        identity_provider = identity_from_url_provider_class.new(authentication_parameters)
+        puts "url token identity avalible? = #{identity_provider.identity_available?}"
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/identity_provider_factory.rb:21]: this should be remved as well

**Review by orenbm — CHANGES_REQUESTED** [review]
left some comments

File: app/domain/authentication/authn_jwt/identity_provider_factory.rb [hunk 4]
```diff
@@ -1,24 +1,35 @@
...
     # If Identity variable is configured factory return the decoded_token_provider
     # If the Identity variable is not configured and there is account field in url factory returns url provider
     # If the above conditions are not met exception is raised
-    class IdentityProviderFactory
-      attr_accessor :from_url_provider, :from_decoded_token_provider
-
-      def initialize(authentication_parameters)
-        @from_url_provider = Authentication::AuthnJwt::IdentityFromUrlProvider.new(authentication_parameters)
-        @from_decoded_token_provider = Authentication::AuthnJwt::IdentityFromDecodedTokenProvider.new(authentication_parameters)
+    IdentityProviderFactory = CommandClass.new(dependencies: {
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/identity_provider_factory.rb:9]: how about `CreateIdentityProvider`? this fits the command class naming pattern better.

Comment (orenbm) [reply app/domain/authentication/authn_jwt/identity_provider_factory.rb:9]: also - move `dependencies` to a new line like in other command classes

File: app/domain/authentication/authn_jwt/configuration_jwt_generic_vendor.rb [hunk 5]
```diff
@@ -3,7 +3,9 @@ module AuthnJwt
     # Mock JWTConfiguration class to use it to develop other part in the jwt authenticator
     class ConfigurationJWTGenericVendor < ConfigurationInterface
       def self.jwt_identity(authentication_parameters)
-        id_provider = Authentication::AuthnJwt::IdentityProviderFactory.new(authentication_parameters).relevant_id_provider
+        id_provider = Authentication::AuthnJwt::IdentityProviderFactory.new.call(
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/configuration_jwt_generic_vendor.rb:6]: can we change this to `identity_provider` to be consistent?

File: app/domain/authentication/authn_jwt/identity_provider_factory.rb (hunk 4 above)
Comment (tzheleznyak) [reply app/domain/authentication/authn_jwt/identity_provider_factory.rb]: Done

File: app/domain/authentication/authn_jwt/create_identity_provider.rb [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    # Factory for jwt identity providers.
+    # If Identity variable is configured factory return the decoded_token_provider
+    # If the Identity variable is not configured and there is account field in url factory returns url provider
+    # If the above conditions are not met exception is raised
+    CreateIdentityProvider = CommandClass.new(dependencies: {
+      identity_from_url_provider_class: Authentication::AuthnJwt::IdentityFromUrlProvider,
+      identity_from_decoded_token_class: Authentication::AuthnJwt::IdentityFromDecodedTokenProvider,
+      logger: Rails.logger
+    },
+                                              inputs: %i[authentication_parameters]
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/create_identity_provider.rb:14]: fix indentation

File: app/domain/authentication/authn_jwt/create_identity_provider.rb [hunk 7]
```diff
@@ -0,0 +1,38 @@
+require 'command_class'
+
+module Authentication
+  module AuthnJwt
+    # Factory for jwt identity providers.
+    # If Identity variable is configured factory return the decoded_token_provider
+    # If the Identity variable is not configured and there is account field in url factory returns url provider
+    # If the above conditions are not met exception is raised
+    CreateIdentityProvider = CommandClass.new(dependencies: {
```
Comment (orenbm) [inline app/domain/authentication/authn_jwt/create_identity_provider.rb:9]: please move `dependencies` to a new line

**Review by orenbm — APPROVED** [review]
love it!

**Pull request merged**